source/geod24/concurrency.d

```diff
@@ -985,6 +986,161 @@ protected:
         bool notified;
     }
 
+    class Channel (T)
+    {
+        this(ulong size = 0)
+        {
+            this.size = size;
+            m_lock = new Mutex;
+        }
+
+        bool write(T val)
+        {
+            bool notify_peer, fiber_was_blocked = false;
+            ChannelQueueEntry qEnt;
+
+            synchronized (m_lock)
+            {
+                notify_peer = m_rxq.length > 0;
+
+                if (m_closed)
+                {
+                    return false;
+                }
+                else if (m_rxq.length > 0)
+                {
+                    qEnt = m_rxq[0];
+                    m_rxq = m_rxq.remove(0);
+
+                    *(qEnt.pVal) = val;
+                }
+                else if (m_buffer.length < size)
+                {
+                    m_buffer ~= val;
+                }
+                else
+                {
+                    qEnt = new ChannelQueueEntry(&val);
```

Comment:
I think this is wrong. You are passing a pointer to a temporary.

Reply:
It might be best to just keep a copy of the value in `ChannelQueueEntry` rather than a pointer.

Reply:
qEnt can be a scoped variable but I think it needs to be a reference. 
At this point, we know that the current fiber will block. If a fiber closes this channel, it will modify the `success` field of the `ChannelQueueEntry` and that needs to be visible to the blocked fiber. Is there a better Dlang way of doing this?

Reply:
Oh, I misunderstood what you meant. The local `val` will not go out of scope as the Fiber will block until pointer to `val` inside `ChannelQueueEntry` is processed. Having a `T*` eliminates the need of extra copy to and from `qEnt` for both reference and value types

Reply:
In fact, to minimize the copying, I tried to make `val` a `auto ref T` but that didn't compile. Compiler strangely says that "auto can only be used as part of auto ref for template function parameters" which is exactly what I am trying to do.

Reply:
Oh so it waits indefinitely until a `read` call processes this `ChannelQueueEntry`?

Reply:
By the way, to make it possible to use `auto ref` you need to make the function a template. Something like: `bool write()(auto ref T val)`.

Reply:
Yes that is correct, they will be blocked until a corresponding `read` comes or channel is `close`'d.